directories: prevent inode cache fragmentation by orderly verifying data directory contents

lkshminarayanan opened the pull request:
During startup, the contents of the data directory are verified to ensure that they have the right owner and permissions. Verifying all the contents, which includes files that will be read and closed immediately, and files that will be held open for longer durations, together, can lead to memory fragementation in the dentry/inode cache.

Mitigate this by updating the verification in a such way that these two set of files will be verified separately ensuring their separation in the dentry/inode cache.

Fixes https://github.com/scylladb/scylladb/issues/14506

lkshminarayanan commented:
From my manual testing, it looks like this patch doesn't improve the situation. The files are loaded into the dentry cache as soon as we check for permissions in the `sstable_directory::filesystem_components_lister::prepare` function call. A better fix is to probably update the prepare method to check for permissions in two phases. I'll update this PR with the new fix.

denesb commented on sstables/sstable_directory.hh:
```diff
@@ -170,13 +171,13 @@ private:
     future<> process_descriptor(sstables::entry_descriptor desc, process_flags flags);
     void validate(sstables::shared_sstable sst, process_flags flags) const;
-    future<sstables::shared_sstable> load_sstable(sstables::entry_descriptor desc, sstables::sstable_open_config cfg = {}) const;
+    future<sstables::shared_sstable> load_sstable(sstables::entry_descriptor desc, sstables::sstable_open_config cfg = {}, bool skip_loading_data_and_index = false) const;
```
Adding a new flag like this introduces a lot of churn (all callers have to be adjusted). We usually split this out into a separate commit, which just adds the flag and updates all callers, but it doesn't use it in the method yet.

Then we add the logic actually using the flag in the method in a 2nd commit. Lastly, a 3rd commit changes the interested callers to pass the non-default value of the flag. If all callers pass the same thing, than this can be folded into the 1st patch.

This amount of separation might seem superfluous, but it makes the review really easy, because reviewers can just glance over the chore patch (adding the flag) and go straight into reviewing the actually interesting changes.

Since this patch is not that big, I don't think you need to split it, just keep this in mind for future PRs.

lkshminarayanan replied:
Sure. I'll keep that in mind for futures PRs.

denesb commented on sstables/sstable_directory.cc:
```diff
@@ -250,6 +250,13 @@ future<> sstable_directory::system_keyspace_components_lister::prepare(sstable_d
 }
 
+future<> sstable_directory::load_data_and_index(process_flags flags) {
+    auto loaded_sstables = flags.sort_sstables_according_to_owner?_unshared_local_sstables:_unsorted_sstables;
```
Space around `?` and `:`.

lkshminarayanan commented:
@denesb @raphaelsc Thank you for your comments! Unfortunately I had to come up a new fix as my previous PR didn't solve the issue. The entries are loaded into the inode/dentry cache as soon as we check for permissions (implemented as `access` syscall internally), so updating the loading mechanism didn't help much. 

This new fix updates the `main` method where the permissions are first checked. It treats data directory as a special case to check the permissions of its contents in two phases. Please let me know your comments in this.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6118/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7421/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6118/)
#### Build Details:
- Duration: 1 hr 46 min
- Builder: spider3.cloudius-systems.com

denesb reviewed (approved):
LGTM, I only have some administrative comments.

It is fine to have patches which just prepare the ground for the next patch. We don't squash these on merge (referring here to the 2nd patch).

The cover letter should always describe the current state of the PR, don't keep out-of-date description or changelog entries in it. These should go to separate comments below. The reason is that when merged, the top-level comment becomes the merge-commit's commit-log, and this should reflect the state of the branch as it was when merged. Changelogs and responses to review or the evolution of the PR looses meaning after the PR is merged.

@bhalevy please also review.

bhalevy commented on utils/directories.cc:
```diff
@@ -77,21 +77,20 @@ directories::directories(bool developer_mode)
-            std::move(locks.begin(), locks.end(), std::back_inserter(_locks));
-        });
+    std::vector<file_lock> locks;
+    co_await parallel_for_each(dir_set.get_paths(), [this, &locks] (fs::path path) -> future<> {
```
nit: please use coroutine::parallel_for_each. It is slightly more efficient for coroutines.

bhalevy commented on utils/directories.cc:
```diff
@@ -77,21 +77,20 @@ directories::directories(bool developer_mode)
-        }).then([this, &locks] {
-            std::move(locks.begin(), locks.end(), std::back_inserter(_locks));
-        });
+    std::vector<file_lock> locks;
```
nit: it would be nice to reserve the vector ahead of time if we know the dir_set size ahead of time.

bhalevy replied:
Not critical since the dir_set is small...

lkshminarayanan replied:
Done.

bhalevy commented on utils/directories.cc:
```diff
@@ -76,14 +76,14 @@ directories::directories(bool developer_mode)
 { }
 
-future<> directories::create_and_verify(directories::set dir_set) {
+future<> directories::create_and_verify(directories::set dir_set, recursive r) {
```
`r` is not descriptive enough for this parameter.
We use single-letter names for variables that are used in smaller scope, or some we use by convention (or used to use) like `t` for table and `s` for schema.
But let's not add more.

bhalevy replied:
About the patch title and comment.
nit: needs a space between data and directory.
If the patch can be applied as stand alone without breaking anything, it's ok to keep it separate and change the note about squashing, just to mention the next step, to provide the motivation for the patch.

lkshminarayanan replied:
Ack.

bhalevy commented on sstables/sstable_version.cc:
```diff
@@ -14,12 +14,14 @@ namespace sstables {
 const sstring sstable_version_constants::TOC_SUFFIX = "TOC.txt";
 const sstring sstable_version_constants::TEMPORARY_TOC_SUFFIX = "TOC.txt.tmp";
+const sstring sstable_version_constants::INDEX_SUFFIX = "Index.db";
+const sstring sstable_version_constants::DATA_SUFFIX = "Data.db";
```
Nit: this change can be refactored into a standalone patch

lkshminarayanan replied:
Done

bhalevy commented on utils/directories.cc:
```diff
@@ -104,7 +105,10 @@ void verification_error(fs::path path, const char* fstr, Args&&... args) {
-future<> directories::do_verify_owner_and_mode(fs::path path, recursive recurse, int level) {
+// If a FilterFunc is provided, verifies only the paths that return true for the filter.
+template <typename FilterFunc>
+requires std::invocable<FilterFunc, fs::path&> && std::is_same_v<std::invoke_result_t<FilterFunc, fs::path&>, bool>
```
Why a template?  Wouldn't a std::function work just as well? (and be simpler)

lkshminarayanan replied:
Agreed. I have changed it to a std::function.

bhalevy commented on utils/directories.cc:
```diff
@@ -115,6 +119,10 @@ future<> directories::do_verify_owner_and_mode(fs::path path, recursive recurse,
     }
     switch (sd.type) {
     case directory_entry_type::regular: {
+        if (!filterFunc(path)) {
```
nit: coding style: we prefer camel_case over camelCase (and try not mix dufferent coding styles in the same unit)

bhalevy replied:
Also, I think it would better to give it a more descriptive name than filter_func.  Maybe do_verify, so the condition will be clearer to the reader.

lkshminarayanan replied:
Done.

bhalevy commented on utils/directories.cc:
```diff
@@ -140,7 +148,23 @@ future<> directories::do_verify_owner_and_mode(fs::path path, recursive recurse,
 
 future<> directories::verify_owner_and_mode(fs::path path, recursive recursive) {
-    return do_verify_owner_and_mode(std::move(path), recursive, 0);
+    return do_verify_owner_and_mode(std::move(path), recursive, 0, [](fs::path &path){return true;});
```
If the signature would use a std::function parameter it should be easy to define this as the default, for backward compatibility.
and nit: please sprinkle some spaces in the lambda definition so it be more readable (like around the parenthesis and inside the curly braces)

lkshminarayanan replied:
Done. An empty function is now the default and `do_verify_owner_and_mode` now checks for empty function before executing it.

bhalevy commented on utils/directories.cc:
```diff
@@ -140,7 +148,23 @@ future<> directories::do_verify_owner_and_mode(fs::path path, recursive recurse,
+}
+
+future<> directories::verify_owner_and_mode_of_data_dir(directories::set dir_set) {
+    co_await parallel_for_each(dir_set.get_paths(), [](const auto &path) -> future<> {
```
nit: coroutine::parallel_for_each

bhalevy replied:
nit: coding style: space around parenthesis

bhalevy commented on utils/directories.cc:
```diff
@@ -140,7 +148,23 @@ future<> directories::do_verify_owner_and_mode(fs::path path, recursive recurse,
+future<> directories::verify_owner_and_mode_of_data_dir(directories::set dir_set) {
+    co_await parallel_for_each(dir_set.get_paths(), [](const auto &path) -> future<> {
+        co_await do_verify_owner_and_mode(std::move(path), recursive::yes, 0, [](fs::path &path) {
+            const sstring &file_name = path.filename().string();
```
nit: coding style: & sticks to type.
also, you can use auto&

And also elsewhere we use native() to get to the path in a native format string  rather than string()

lkshminarayanan replied:
Done.

bhalevy commented on utils/directories.cc:
```diff
@@ -140,7 +148,23 @@ future<> directories::do_verify_owner_and_mode(fs::path path, recursive recurse,
+        });
+    });
+
+    co_await parallel_for_each(dir_set.get_paths(), [](const auto &path) -> future<> {
```
nit: coroutine::parallel_for_each, space around parens

lkshminarayanan replied:
Done.

bhalevy commented on utils/directories.cc:
```diff
@@ -140,7 +148,23 @@ future<> directories::do_verify_owner_and_mode(fs::path path, recursive recurse,
+    co_await parallel_for_each(dir_set.get_paths(), [](const auto &path) -> future<> {
+        co_await do_verify_owner_and_mode(std::move(path), recursive::yes, 0, [](fs::path &path) {
+            const sstring &file_name = path.filename().string();
+            return !(file_name.ends_with(sstables::sstable_version_constants::DATA_SUFFIX) || file_name.ends_with(sstables::sstable_version_constants::INDEX_SUFFIX));
```
this block us similar to the above that I wonder if we can encapsulate it in a reusable function

lkshminarayanan replied:
I have now updated this code to run in a loop to achieve the same results.

avikivity commented:
How did you test it?

lkshminarayanan commented:
@bhalevy @denesb Thank you for your reviews! I have updated the PR with a new set of commits addressing the review comments. Please let me know if this looks okay.

lkshminarayanan commented:
> How did you test it?

@avikivity I was only able to test the timing of the entries getting loaded into the dentry/inode cache using `slabtop`. Using this tool, I am able to confirm that `verify_owner_and_mode_of_data_dir` does populate the inode/dentry cache. But, apart from that, I am unable to find ways to look at the actual slabs themselves.

avikivity commented:
> > How did you test it?
> 
> @avikivity I was only able to test the timing of the entries getting loaded into the dentry/inode cache using `slabtop`. Using this tool, I am able to confirm that `verify_owner_and_mode_of_data_dir` does populate the inode/dentry cache. But, apart from that, I am unable to find ways to look at the actual slabs themselves.

Try this:

1. echo 3 > /proc/sys/vm/drop_caches, look at slab sizes (slabtop)
2. populate a few gigabytes using LeveledCompactionStrategy with sstable_size_in_mb = 1, giving a few thousand sstables
3. look at slab sizes again
4. drop_caches again (see it return to original size)
5. start the node, let it populate
6. look at slab caches, should grow by total number of files in data dir
7. drop_caches (while the node is up). look at slab caches, should show just 2 files/sstable on top of base load (in step 1)


You can repeat 6, 7 with and without the patch. Without the patch drop_caches should not materially shrink memory in use.

avikivity commented:
Please adjust the subject to reflect the goal too, we're not changing the verification order for no reason.

scylladb-promoter commented:
#### :red_circle: CI State: FAILURE

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6182/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6182/)
:x: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7483/)
:x: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6182/)
#### Failed Tests (163/24108):
- [test_replace_node_after_cluster_restart](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6182/testReport/junit/repair_based_node_operations_test/TestRepairBasedNodeOperations/Tests___dtest___test_replace_node_after_cluster_restart) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_replace_node_after_cluster_restart)
- [test_basic](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6182/testReport/junit/%28root%29/test_basic/Tests___Unit_Tests___test_basic) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_basic)
- [test_garbage_collect](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6182/testReport/junit/%28root%29/test_basic/Tests___Unit_Tests___test_garbage_collect) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_garbage_collect)
- [test_memtable_flush_retries](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6182/testReport/junit/%28root%29/test_basic/Tests___Unit_Tests___test_memtable_flush_retries) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_memtable_flush_retries)
- [test_basic](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6182/testReport/junit/%28root%29/test_basic/Tests___Unit_Tests___test_basic_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_basic)
- [test_garbage_collect](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6182/testReport/junit/%28root%29/test_basic/Tests___Unit_Tests___test_garbage_collect_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_garbage_collect)
- [test_memtable_flush_retries](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6182/testReport/junit/%28root%29/test_basic/Tests___Unit_Tests___test_memtable_flush_retries_2) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_memtable_flush_retries)
- [test_basic](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6182/testReport/junit/%28root%29/test_basic/Tests___Unit_Tests___test_basic_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_basic)
- [test_garbage_collect](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6182/testReport/junit/%28root%29/test_basic/Tests___Unit_Tests___test_garbage_collect_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_garbage_collect)
- [test_memtable_flush_retries](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6182/testReport/junit/%28root%29/test_basic/Tests___Unit_Tests___test_memtable_flush_retries_3) [:mag:](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_memtable_flush_retries)
#### Build Details:
- Duration: 1 hr 33 min
- Builder: spider8.cloudius-systems.com

lkshminarayanan commented:
Hi @denesb @bhalevy, I have updated the PR again. The previous version didn't work as expected as the very first line of [do_verify_owner_and_mode()](https://github.com/scylladb/scylladb/blob/0cbf8f75f0163898cd6359b843f8c9d3237177f3/utils/directories.cc#L109) does a `file_stat` call and thus the inode entry gets into the cache at that point. Filtering out paths after turned out to be pointless. 

The newer version of the PR applies this filter before any of these system calls and succeeds in loading the files in the specific order.

lkshminarayanan commented:
> > > How did you test it?
> > 
> > 
> > @avikivity I was only able to test the timing of the entries getting loaded into the dentry/inode cache using `slabtop`. Using this tool, I am able to confirm that `verify_owner_and_mode_of_data_dir` does populate the inode/dentry cache. But, apart from that, I am unable to find ways to look at the actual slabs themselves.
> 
> Try this:
> 
> 1. echo 3 > /proc/sys/vm/drop_caches, look at slab sizes (slabtop)
> 2. populate a few gigabytes using LeveledCompactionStrategy with sstable_size_in_mb = 1, giving a few thousand sstables
> 3. look at slab sizes again
> 4. drop_caches again (see it return to original size)
> 5. start the node, let it populate
> 6. look at slab caches, should grow by total number of files in data dir
> 7. drop_caches (while the node is up). look at slab caches, should show just 2 files/sstable on top of base load (in step 1)
> 
> You can repeat 6, 7 with and without the patch. Without the patch drop_caches should not materially shrink memory in use.

Thank you! I was able to test the PR and the number of slabs does get reduced with the latest version of PR.

One sample from my tests :
Without the patch, total number of slabs :
1. After drop_caches but before populating the keyspaces  : 14701
2. After populating the keyspaces : 24316
3. After drop_caches : 22305

With the patch, total number of slabs :
1. After drop_caches but before populating the keyspaces  : 14401
2. After populating the keyspaces : 24212
3. After drop_caches : 16355

I can see a similar drop in the slabs used by the dentry and inode caches as well.

denesb commented:
> #### 🔴 CI State: FAILURE
> 
> ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6215/) ✅ - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6215/) ❌ - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6215/) ✅ - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7514/)
> #### Failed Tests (3/24108):
> 
>     * [test_tablet_cleanup](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6215/testReport/junit/%28root%29/test_tablets/Tests___Unit_Tests___test_tablet_cleanup) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_tablet_cleanup)
> 
>     * [topology_experimental_raft.test_tablets.release.6](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6215/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___topology_experimental_raft_test_tablets_release_6) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+topology_experimental_raft.test_tablets.release.6)

https://github.com/scylladb/scylladb/issues/14746 (fixed already)
> 
>     * [test_token_group_based_splitting_mutation_writer](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6215/testReport/junit/boost/mutation_writer_test/test_token_group_based_splitting_mutation_writer) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_token_group_based_splitting_mutation_writer)
> 

https://github.com/scylladb/scylladb/issues/17076

> 
> #### Build Details:
> 
>     * Duration: 1 hr 55 min
> 
>     * Builder: spider5.cloudius-systems.com

@lkshminarayanan please rebase on current master, there are conflicts. Also, some of the above CI failures have been dealt with already.

lkshminarayanan commented:
@denesb I have now rebased the PR on top of master. A recent PR moved the verification code from main into directories class. So, I have moved my fix into there as well.

denesb commented:
> #### 🔴 CI State: FAILURE
> 
> ✅ - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6234/) ✅ - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6234/) ❌ - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6234/) ✅ - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7533/)
> #### Failed Tests (2/24144):
> 
>     * [test_login_message_after_half_of_the_cluster_is_down](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6234/testReport/junit/%28root%29/test_password_login_message/Tests___Unit_Tests___test_login_message_after_half_of_the_cluster_is_down) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+test_login_message_after_half_of_the_cluster_is_down)
> 
>     * [auth_cluster.test_password_login_message.debug.1](https://jenkins.scylladb.com//job/scylla-master/job/scylla-ci/6234/testReport/junit/%28root%29/non-boost%20tests/Tests___Unit_Tests___auth_cluster_test_password_login_message_debug_1) [🔍](https://github.com/scylladb/scylladb/issues?q=is:issue+is:open+auth_cluster.test_password_login_message.debug.1)
> 
> 
> #### Build Details:
> 
>     * Duration: 1 hr 39 min
> 
>     * Builder: spider7.cloudius-systems.com

https://github.com/scylladb/scylladb/issues/16821

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6241/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6241/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7539/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6241/)
#### Build Details:
- Duration: 1 hr 34 min
- Builder: spider7.cloudius-systems.com

denesb commented:
@avikivity @bhalevy please re-review.

lkshminarayanan commented:
Rebased on top of latest master.

scylladb-promoter commented:
#### :green_circle: CI State: SUCCESS

:white_check_mark: - [Build](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6276/)
:white_check_mark: - [Container Test](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6276/)
:white_check_mark: - [dtest](https://jenkins.scylladb.com/job/scylla-master/job/gating-dtest-release/7575/)
:white_check_mark: - [Unit Tests](https://jenkins.scylladb.com/job/scylla-master/job/scylla-ci/6276/)
#### Build Details:
- Duration: 1 hr 57 min
- Builder: spider4.cloudius-systems.com
